File: server/bot.js
import "dotenv/config";
import TelegramBot from "node-telegram-bot-api";

const TOKEN = process.env.TELEGRAM_BOT_TOKEN || "YOUR_TELEGRAM_BOT_TOKEN";
const SUPABASE_URL = process.env.SUPABASE_URL || "https://YOUR_SUPABASE_URL";
const SUPABASE_SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || "YOUR_SUPABASE_SERVICE_ROLE_KEY";
const WEBAPP_URL = process.env.WEBAPP_URL || "https://YOUR_WEBAPP_URL";

if (TOKEN.includes("YOUR_")) {
  console.warn("Set TELEGRAM_BOT_TOKEN before запуск");
}

const bot = new TelegramBot(TOKEN, { polling: true });

async function upsertUser(user) {
  if (SUPABASE_URL.includes("YOUR_")) return;
  const payload = {
    id: String(user.id),
    username: user.username || null,
    display_name: user.first_name || "Игрок",
    photo_url: user.photo_url || null,
    badges: ["Новичок"],
    season: "Зима 2025",
    achievements: "0/30",
  };

  await fetch(`${SUPABASE_URL}/rest/v1/users`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      apikey: SUPABASE_SERVICE_ROLE_KEY,
      Authorization: `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
      Prefer: "resolution=merge-duplicates",
    },
    body: JSON.stringify(payload),
  });
}

bot.onText(/\/(start|menu)/, async (msg) => {
  const chatId = msg.chat.id;
  if (msg.from) {
    try {
      await upsertUser(msg.from);
    } catch (err) {
      console.error("Supabase error", err.message);
    }
  }

  bot.sendMessage(chatId, "Открыть Снежный Кликер", {
    reply_markup: {
      inline_keyboard: [
        [{ text: "Открыть Mini App", web_app: { url: WEBAPP_URL } }],
      ],
    },
  });
});

bot.on("message", (msg) => {
  if (msg.text && msg.text.startsWith("/")) return;
  bot.sendMessage(msg.chat.id, "Нажми кнопку, чтобы открыть игру.");
});

File: server/bot.py
from __future__ import annotations

import os
import asyncio

from aiogram import Bot, Dispatcher, types
from aiogram.filters import Command

TOKEN = os.getenv("TELEGRAM_BOT_TOKEN", "YOUR_TELEGRAM_BOT_TOKEN")
SUPABASE_URL = os.getenv("SUPABASE_URL", "https://YOUR_SUPABASE_URL")
SUPABASE_ANON_KEY = os.getenv("SUPABASE_ANON_KEY", "YOUR_SUPABASE_ANON_KEY")


async def upsert_user(user: types.User) -> None:
    if "YOUR_" in SUPABASE_URL or "YOUR_" in SUPABASE_ANON_KEY:
        return

    payload = {
        "id": str(user.id),
        "username": user.username,
        "display_name": user.first_name or "Игрок",
        "photo_url": None,
        "badges": ["Новичок"],
        "season": "Зима 2025",
        "achievements": "0/30",
    }

    headers = {
        "Content-Type": "application/json",
        "apikey": SUPABASE_ANON_KEY,
        "Authorization": f"Bearer {SUPABASE_ANON_KEY}",
        "Prefer": "resolution=merge-duplicates",
    }

    import aiohttp

    async with aiohttp.ClientSession() as session:
        async with session.post(f"{SUPABASE_URL}/rest/v1/users", json=payload, headers=headers) as resp:
            if resp.status >= 400:
                text = await resp.text()
                print("Supabase error", resp.status, text)


async def start_handler(message: types.Message) -> None:
    if message.from_user:
        await upsert_user(message.from_user)
    await message.answer("Открыть Снежный Кликер через меню бота.")


async def fallback_handler(message: types.Message) -> None:
    if message.text and message.text.startswith("/"):
        return
    await message.answer("Нажми кнопку, чтобы открыть игру.")


async def main() -> None:
    if "YOUR_" in TOKEN:
        print("Set TELEGRAM_BOT_TOKEN before запуск")

    bot = Bot(token=TOKEN)
    dp = Dispatcher()

    dp.message.register(start_handler, Command("start", "menu"))
    dp.message.register(fallback_handler)

    await dp.start_polling(bot)


if __name__ == "__main__":
    asyncio.run(main())
